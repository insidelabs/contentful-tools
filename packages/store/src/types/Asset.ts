export interface Asset {
    __typename: 'Asset';
    __id: string;
    title?: string;
    description?: string;
    file: {
        url: string;
        contentType: string;
        fileName: string;
        details: {
            size: number;
            image?: {
                width: number;
                height: number;
            };
        };
    };
}
