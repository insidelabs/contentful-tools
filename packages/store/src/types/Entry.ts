export interface Entry {
    __typename: string;
    __id: string;
}
