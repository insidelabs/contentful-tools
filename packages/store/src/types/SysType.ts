export type SysType =
    | 'Asset'
    | 'ContentType'
    | 'DeletedAsset'
    | 'DeletedEntry'
    | 'Entry'
    | 'Environment'
    | 'Link'
    | 'Space';
