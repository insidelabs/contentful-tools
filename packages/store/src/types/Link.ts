import { SysType } from './SysType';

export interface AssetLink {
    sys: {
        type: 'Link';
        linkType: 'Asset';
        id: string;
    };
}

export interface EntryLink {
    sys: {
        type: 'Link';
        linkType: 'Entry';
        id: string;
    };
}

export interface SpaceLink {
    sys: {
        type: 'Link';
        linkType: 'Space';
        id: string;
    };
}

export interface EnvironmentLink {
    sys: {
        type: 'Link';
        linkType: 'Environment';
        id: string;
    };
}

export interface ContentTypeLink {
    sys: {
        type: 'Link';
        linkType: 'ContentType';
        id: string;
    };
}

export interface Link {
    sys: {
        type: 'Link';
        linkType: SysType;
        id: string;
    };
}
