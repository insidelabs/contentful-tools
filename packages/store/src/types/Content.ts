import { Link } from './Link';
import { SysType } from './SysType';

export namespace Content {
    export interface Asset {
        sys: {
            id: string;
            type: SysType.Asset;
            space: Link.Space;
            environment: Link.Environment;
            createdAt: string;
            updatedAt: string;
            revision: number;
        };
        fields: AssetFields;
    }

    export interface AssetFields {
        title?: string;
        description?: string;
        file: {
            url: string;
            details: {
                size: number;
                image?: {
                    width: number;
                    height: number;
                };
            };
            fileName: string;
            contentType: string;
        };
    }

    export interface Entry<ContentTypeId extends string = string> {
        sys: {
            id: string;
            type: SysType.Entry;
            space: Link.Space;
            environment: Link.Environment;
            contentType: Link.ContentType<ContentTypeId>;
            createdAt: string;
            updatedAt: string;
            revision: number;
        };
        fields: EntryFields;
    }

    export type EntryFields = {
        [key: string]: Link.Asset | Link.Asset[] | Link.Entry | Link.Entry[] | unknown;
    };
}
