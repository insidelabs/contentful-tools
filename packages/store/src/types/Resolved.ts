import { Content } from './Content';
import { Link } from './Link';

export namespace Resolved {
    export type Asset<A extends Content.Asset = Content.Asset> = A['fields'] & {
        __id: A['sys']['id'];
    };

    export type Entry<E extends Content.Entry = Content.Entry> = EntryFields<E['fields']> & {
        __typename: E['sys']['contentType']['sys']['id'];
        __id: E['sys']['id'];
    };

    type EntryFields<F extends Content.EntryFields> = RequiredFields<F> & OptionalFields<F>;

    type RequiredFields<F extends Content.EntryFields> = {
        readonly [K in RequiredKeys<F>]-?: F[K] extends Link.Asset
            ? Asset<Content.Asset>
            : F[K] extends Link.Asset[]
            ? Asset<Content.Asset>[]
            : F[K] extends Link.Entry<infer L>
            ? Entry<L>
            : F[K] extends Link.Entry<infer L>[]
            ? Entry<L>[]
            : F[K];
    };

    type OptionalFields<F extends Content.EntryFields> = {
        readonly [K in OptionalKeys<F>]+?: NonOptional<F[K]> extends Link.Asset
            ? Asset<Content.Asset>
            : NonOptional<F[K]> extends Link.Asset[]
            ? Asset<Content.Asset>[]
            : NonOptional<F[K]> extends Link.Entry<infer L>
            ? Entry<L>
            : NonOptional<F[K]> extends Link.Entry<infer L>[]
            ? Entry<L>[]
            : F[K];
    };

    type RequiredKeys<T> = { [K in keyof T]-?: undefined extends T[K] ? never : K }[keyof T];
    type OptionalKeys<T> = { [K in keyof T]-?: undefined extends T[K] ? K : never }[keyof T];

    type NonOptional<T> = T extends undefined ? never : T;
}
