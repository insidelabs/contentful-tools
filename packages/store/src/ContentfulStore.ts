import createDebugger, { Debugger } from 'debug';
import { ContentfulClientApi } from 'contentful';
import { Content } from './types/Content';
import { Resolved } from './types/Resolved';
import { Sync } from './types/Sync';
import { Util } from './types/Util';
import { isAssetLink, isAssetLinkArray, isEntryLink, isEntryLinkArray } from './guards';

export namespace ContentfulStore {
    export interface Config<BaseLocale extends string, ExtraLocales extends string> {
        client: ContentfulClientApi;
        spaceId: string;
        locales: [BaseLocale, ...ExtraLocales[]];
        handleError?: ErrorHandler;
        autoSync?: {
            minInterval: number;
        };
    }

    export type ErrorHandler = (error: Error) => void;
}

namespace Internals {
    export interface AutoSync {
        readonly enabled: boolean;
        readonly minInterval: number;
        requestCount: number;
        timeout: NodeJS.Timeout | null;
    }

    export type Assets = LocaleMapped<AssetMap>;
    export type Entries = LocaleMapped<EntryMap>;

    type AssetMap = Map<string, Content.Asset>;
    type EntryMap = Map<string, Resolved.Entry>;

    type LocaleMapped<T> = { [locale in string]: T };
}

export class ContentfulStore<BaseLocale extends string, ExtraLocales extends string> {
    private readonly client: ContentfulClientApi;
    private readonly locales: [BaseLocale, ...ExtraLocales[]];

    private readonly debug: Debugger;
    private readonly handleError: ContentfulStore.ErrorHandler;

    private readonly autoSync: Internals.AutoSync;

    private readonly assets: Internals.Assets = {};
    private readonly entries: Internals.Entries = {};

    private syncToken?: string;

    constructor({
        client,
        spaceId,
        locales,
        handleError,
        autoSync,
    }: ContentfulStore.Config<BaseLocale, ExtraLocales>) {
        this.client = client;
        this.locales = locales;

        this.debug = createDebugger(`@contentful-tools/store:${spaceId}`);

        this.handleError =
            handleError ||
            ((error: Error) => {
                console.error(`Error in ContentfulStore (${spaceId})`, error);
            });

        this.autoSync = {
            enabled: autoSync != null,
            minInterval: autoSync ? autoSync.minInterval : 0,
            requestCount: 0,
            timeout: null,
        };

        for (const locale of locales) {
            this.assets[locale] = new Map();
            this.entries[locale] = new Map();
        }
    }

    private get baseLocale(): BaseLocale {
        return this.locales[0];
    }

    public getAsset(
        id: string,
        locale: BaseLocale | ExtraLocales = this.baseLocale,
    ): Content.Asset | null {
        this.onContentAccess();
        return this.assets[locale].get(id) || null;
    }

    public getAssets(locale: BaseLocale | ExtraLocales = this.baseLocale): Content.Asset[] {
        this.onContentAccess();
        return Array.from(this.assets[locale].values());
    }

    public getEntry<E extends Content.Entry = Content.Entry>(
        id: string,
        locale: BaseLocale | ExtraLocales = this.baseLocale,
        contentTypeId?: Util.GetContentTypeId<E>,
    ): Resolved.Entry<E> | null {
        this.onContentAccess();
        const entry = this.entries[locale].get(id);
        return entry && (entry.sys.contentType.sys.id === contentTypeId || !contentTypeId)
            ? (entry as Resolved.Entry<E>)
            : null;
    }

    public getEntryByFieldValue<
        E extends Content.Entry,
        F extends keyof E['fields'] = keyof E['fields'],
        V extends E['fields'][F] = E['fields'][F],
    >(
        field: F,
        value: V,
        locale: BaseLocale | ExtraLocales = this.baseLocale,
        contentTypeId?: Util.GetContentTypeId<E>,
    ): Resolved.Entry<E> | null {
        this.onContentAccess();
        const entries = this.getEntries<E>(locale, contentTypeId);
        const entry = entries.find(e => (e.fields as any)[field] === value);
        return entry || null;
    }

    public getEntries<E extends Content.Entry = Content.Entry>(
        locale: BaseLocale | ExtraLocales = this.baseLocale,
        contentTypeId?: Util.GetContentTypeId<E>,
    ): Resolved.Entry<E>[] {
        this.onContentAccess();
        const entries = Array.from(this.entries[locale].values());
        return (contentTypeId
            ? entries.filter(entry => entry.sys.contentType.sys.id === contentTypeId)
            : entries) as Resolved.Entry<E>[];
    }

    public async sync(): Promise<void> {
        const query: Sync.Query = { resolveLinks: false };

        if (this.syncToken) query.nextSyncToken = this.syncToken;
        else query.initial = true;

        const result = await this.client.sync(query);

        const {
            assets,
            entries,
            deletedAssets,
            deletedEntries,
            nextSyncToken,
        } = result.toPlainObject() as Sync.Result<BaseLocale, ExtraLocales>;

        this.debug(`Synced ${assets.length} assets`);
        this.debug(`Synced ${entries.length} entries`);
        this.debug(`Deleted ${deletedAssets.length} assets`);
        this.debug(`Deleted ${deletedEntries.length} entries`);

        this.syncToken = nextSyncToken;

        for (const locale of this.locales) {
            for (const asset of assets) {
                this.assets[locale].set(asset.sys.id, this.processSyncAsset(asset, locale));
            }

            for (const entry of entries) {
                this.entries[locale].set(entry.sys.id, this.processSyncEntry(entry, locale));
            }

            for (const asset of deletedAssets) this.assets[locale].delete(asset.sys.id);
            for (const entry of deletedEntries) this.entries[locale].delete(entry.sys.id);
        }
    }

    private processSyncAsset(
        asset: Sync.Asset<BaseLocale, ExtraLocales>,
        locale: BaseLocale | ExtraLocales,
    ): Content.Asset {
        return {
            sys: asset.sys,
            fields: this.processFieldsForLocale(locale, asset.fields),
        };
    }

    private processSyncEntry(
        entry: Sync.Entry<BaseLocale, ExtraLocales>,
        locale: BaseLocale | ExtraLocales,
    ): Resolved.Entry {
        return this.createResolvedEntry(locale, {
            sys: entry.sys,
            fields: this.processFieldsForLocale(locale, entry.fields),
        });
    }

    private processFieldsForLocale<F extends { [key: string]: any }>(
        locale: BaseLocale | ExtraLocales,
        fields: Sync.Fields<F, BaseLocale, ExtraLocales>,
    ): F {
        return Object.assign(
            {},
            ...Object.keys(fields).map(key => ({
                [key]: fields[key][locale] || fields[key][this.baseLocale],
            })),
        );
    }

    private createResolvedEntry<E extends Content.Entry>(
        locale: BaseLocale | ExtraLocales,
        entry: E,
    ): Resolved.Entry<E> {
        const fields: { [key: string]: unknown } = {};

        for (const [key, value] of Object.entries(entry.fields)) {
            const descriptor: PropertyDescriptor = {
                enumerable: true,
                configurable: false,
            };

            if (isAssetLink(value)) {
                descriptor.get = () => this.assets[locale].get(value.sys.id);
            } else if (isAssetLinkArray(value)) {
                descriptor.get = () => value.map(link => this.assets[locale].get(link.sys.id));
            } else if (isEntryLink(value)) {
                descriptor.get = () => this.entries[locale].get(value.sys.id);
            } else if (isEntryLinkArray(value)) {
                descriptor.get = () => value.map(link => this.entries[locale].get(link.sys.id));
            } else {
                descriptor.value = value;
                descriptor.writable = false;
            }

            Object.defineProperty(fields, key, descriptor);
        }

        return {
            sys: entry.sys,
            fields: fields,
        } as Resolved.Entry<E>;
    }

    private onContentAccess() {
        if (!this.syncToken) {
            this.handleError(Error('Content accessed without initial sync'));
        }

        if (!this.autoSync.enabled) return;

        this.autoSync.requestCount++;
        if (this.autoSync.requestCount > 1) return;

        this.sync().catch(this.handleError);

        this.autoSync.timeout = setTimeout(() => {
            if (this.autoSync.requestCount > 1) this.sync().catch(this.handleError);
            this.autoSync.requestCount = 0;
        }, this.autoSync.minInterval);
    }
}
