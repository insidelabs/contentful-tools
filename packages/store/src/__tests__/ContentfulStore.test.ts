import { ContentfulStore, ContentfulStoreConfig } from '../ContentfulStore';
import { client } from './__stubs__/contentful.stub';
import { Asset } from '../types/Asset';
import SpyInstance = jest.SpyInstance;

type BaseLocale = 'en';
type ExtraLocales = 'de';

interface Widget {
    __typename: 'Widget';
    __id: string;
    name: string;
    description?: string;
    child: Widget;
    relatives?: Widget[];
    image: Asset;
    images?: Asset[];
}

describe('ContentfulStore', () => {
    const autoSync = { minInterval: 10000 };
    let store: ContentfulStore<BaseLocale, ExtraLocales>;

    function createStore(config: Partial<ContentfulStoreConfig<BaseLocale, ExtraLocales>> = {}) {
        return new ContentfulStore({
            ...config,
            client,
            spaceId: 'widget-space',
            locales: ['en', 'de'],
        });
    }

    describe('store', () => {
        beforeEach(() => {
            store = createStore();
        });

        it('should load', async () => {
            await store.sync();
            expect(store.getEntries()).toMatchSnapshot();
            expect(store.getAssets()).toMatchSnapshot();
        });

        it('should sync', async () => {
            await store.sync();
            await store.sync();
            expect(store.getEntries()).toMatchSnapshot();
            expect(store.getAssets()).toMatchSnapshot();
        });
    });

    describe('auto-sync', () => {
        let spy: SpyInstance<typeof client.sync>;

        beforeAll(() => {
            jest.useFakeTimers();
        });

        beforeEach(async () => {
            store = createStore({ autoSync });
            await store.sync();
            spy = (jest.spyOn(client, 'sync') as unknown) as SpyInstance<typeof client.sync>;
        });

        afterEach(() => {
            spy.mockRestore();
            jest.clearAllTimers();
        });

        afterAll(() => {
            jest.useRealTimers();
        });

        it('should auto-sync on getting an asset', () => {
            store.getAsset('bargis');
            expect(spy).toHaveBeenCalledTimes(1);
        });

        it('should auto-sync on getting assets', () => {
            store.getAssets();
            expect(spy).toHaveBeenCalledTimes(1);
        });

        it('should auto-sync on getting an entry', () => {
            store.getEntry('foo');
            expect(spy).toHaveBeenCalledTimes(1);
        });

        it('should auto-sync on getting an entry by field value', () => {
            store.getEntryByFieldValue<Widget>('name', 'Foo');
        });

        it('should auto-sync on getting entries', () => {
            store.getEntries();
            expect(spy).toHaveBeenCalledTimes(1);
        });

        it('should throttle auto-sync', () => {
            store.getEntries();
            store.getEntries();
            expect(spy).toHaveBeenCalledTimes(1);
        });

        it('should auto-sync on leading & trailing edge of interval', () => {
            store.getEntries();
            store.getEntries();
            expect(spy).toHaveBeenCalledTimes(1);
            jest.advanceTimersByTime(autoSync.minInterval);
            expect(spy).toHaveBeenCalledTimes(2);
        });

        it('should release auto-sync throttle after interval', () => {
            store.getEntries();
            expect(spy).toHaveBeenCalledTimes(1);

            jest.advanceTimersByTime(autoSync.minInterval);

            store.getEntries();
            expect(spy).toHaveBeenCalledTimes(2);
        });

        it('should only auto-sync if enabled', async () => {
            store = createStore();
            await store.sync();
            spy.mockReset();
            store.getAsset('bargis');
            expect(spy).not.toBeCalled();
        });
    });

    describe('content', () => {
        beforeAll(async () => {
            store = createStore();
            await store.sync();
        });

        it('should return an asset', () => {
            const asset = store.getAsset('bargis');
            expect(asset!.__id).toMatchInlineSnapshot(`"bargis"`);
        });

        it('should return null for a missing asset', () => {
            const asset = store.getAsset('missing');
            expect(asset).toBeNull();
        });

        it('should return an entry', () => {
            const entry = store.getEntry<Widget>('foo');
            expect(entry!.__id).toMatchInlineSnapshot(`"foo"`);
        });

        it('should return an entry by field value', () => {
            const entry = store.getEntryByFieldValue<Widget, 'name'>('name', 'Foo');
            expect(entry!.__id).toMatchInlineSnapshot(`"foo"`);
        });

        it('should return null for a missing entry', () => {
            const entry = store.getEntry('missing');
            expect(entry).toBeNull();
        });

        it('should return entries of a specific content type', () => {
            const entries = store.getEntries('en', 'Gadget');
            expect(entries).toHaveLength(1);
            expect(entries[0].__id).toMatchInlineSnapshot(`"doodah"`);
        });
    });

    describe('links', () => {
        let foo: Widget;

        beforeAll(async () => {
            store = await createStore();
            await store.sync();
        });

        beforeEach(() => {
            foo = store.getEntry<Widget>('foo')!;
        });

        it('should resolve a linked asset', () => {
            const asset = foo.image;
            expect(asset.__id).toMatchInlineSnapshot(`"bargis"`);
        });

        it('should resolve a linked entry', () => {
            const child = foo.child;
            expect(child.__id).toMatchInlineSnapshot(`"bar"`);
        });

        it('should resolve a deeply-linked entry', () => {
            const grandchild = foo.child.child;
            expect(grandchild.__id).toMatchInlineSnapshot(`"baz"`);
        });

        it('should resolve a circularly-linked entry', () => {
            const qux = store.getEntry<Widget>('qux')!;
            const self = qux.child;
            expect(self.__id).toMatchInlineSnapshot(`"qux"`);
        });

        it('should resolve an array of linked assets', () => {
            const assets = foo.images!;
            expect(assets.map(asset => asset.__id)).toMatchInlineSnapshot(`
                Array [
                  "bargis",
                  "padasjoki",
                ]
            `);
        });

        it('should resolve an array of linked entries', () => {
            const entries = foo.relatives!;
            expect(entries.map(entry => entry.__id)).toMatchInlineSnapshot(`
                Array [
                  "baz",
                  "qux",
                ]
            `);
        });
    });

    describe('errors', () => {
        it('should log error on content access if not yet loaded', () => {
            const spy = jest.spyOn(global.console, 'error').mockImplementation(() => {});
            store = createStore();
            store.getEntries();
            expect(spy).toHaveBeenCalledTimes(1);
            expect(spy.mock.calls[0]).toMatchInlineSnapshot(`
                Array [
                  "Error in ContentfulStore (widget-space)",
                  [Error: Content accessed without initial sync],
                ]
            `);
        });

        it('should pass error to a custom handler', () => {
            const spy = jest.fn();
            store = createStore({ handleError: spy });
            store.getEntries();
            expect(spy).toHaveBeenCalledTimes(1);
            spy.mockRestore();
        });
    });
});
