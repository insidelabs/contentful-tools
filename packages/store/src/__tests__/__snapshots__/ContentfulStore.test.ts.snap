// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ContentfulStore store should load 1`] = `
Array [
  Object {
    "__id": "foo",
    "__typename": "Widget",
    "child": Object {
      "__id": "bar",
      "__typename": "Widget",
      "child": Object {
        "__id": "baz",
        "__typename": "Widget",
        "child": Object {
          "__id": "qux",
          "__typename": "Widget",
          "child": [Circular],
          "image": Object {
            "__id": "bargis",
            "__typename": "Asset",
            "file": Object {
              "contentType": "image/jpeg",
              "details": Object {
                "image": Object {
                  "height": 2828,
                  "width": 4242,
                },
                "size": 3996687,
              },
              "fileName": "bargis.jpg",
              "url": "//images.ctfassets.net/abc/bargis/456/bargis.jpg",
            },
            "title": "Bargis",
          },
          "name": "Qux",
        },
        "image": Object {
          "__id": "bargis",
          "__typename": "Asset",
          "file": Object {
            "contentType": "image/jpeg",
            "details": Object {
              "image": Object {
                "height": 2828,
                "width": 4242,
              },
              "size": 3996687,
            },
            "fileName": "bargis.jpg",
            "url": "//images.ctfassets.net/abc/bargis/456/bargis.jpg",
          },
          "title": "Bargis",
        },
        "name": "Baz",
      },
      "image": Object {
        "__id": "padasjoki",
        "__typename": "Asset",
        "file": Object {
          "contentType": "image/jpeg",
          "details": Object {
            "image": Object {
              "height": 3000,
              "width": 4000,
            },
            "size": 5080466,
          },
          "fileName": "padasjoki.jpg",
          "url": "//images.ctfassets.net/abc/padasjoki/123/padasjoki.jpg",
        },
        "title": "padasjoki",
      },
      "name": "Bar",
      "relatives": Array [
        [Circular],
      ],
    },
    "description": "An English Foo",
    "image": Object {
      "__id": "bargis",
      "__typename": "Asset",
      "file": Object {
        "contentType": "image/jpeg",
        "details": Object {
          "image": Object {
            "height": 2828,
            "width": 4242,
          },
          "size": 3996687,
        },
        "fileName": "bargis.jpg",
        "url": "//images.ctfassets.net/abc/bargis/456/bargis.jpg",
      },
      "title": "Bargis",
    },
    "images": Array [
      Object {
        "__id": "bargis",
        "__typename": "Asset",
        "file": Object {
          "contentType": "image/jpeg",
          "details": Object {
            "image": Object {
              "height": 2828,
              "width": 4242,
            },
            "size": 3996687,
          },
          "fileName": "bargis.jpg",
          "url": "//images.ctfassets.net/abc/bargis/456/bargis.jpg",
        },
        "title": "Bargis",
      },
      Object {
        "__id": "padasjoki",
        "__typename": "Asset",
        "file": Object {
          "contentType": "image/jpeg",
          "details": Object {
            "image": Object {
              "height": 3000,
              "width": 4000,
            },
            "size": 5080466,
          },
          "fileName": "padasjoki.jpg",
          "url": "//images.ctfassets.net/abc/padasjoki/123/padasjoki.jpg",
        },
        "title": "padasjoki",
      },
    ],
    "name": "Foo",
    "relatives": Array [
      Object {
        "__id": "baz",
        "__typename": "Widget",
        "child": Object {
          "__id": "qux",
          "__typename": "Widget",
          "child": [Circular],
          "image": Object {
            "__id": "bargis",
            "__typename": "Asset",
            "file": Object {
              "contentType": "image/jpeg",
              "details": Object {
                "image": Object {
                  "height": 2828,
                  "width": 4242,
                },
                "size": 3996687,
              },
              "fileName": "bargis.jpg",
              "url": "//images.ctfassets.net/abc/bargis/456/bargis.jpg",
            },
            "title": "Bargis",
          },
          "name": "Qux",
        },
        "image": Object {
          "__id": "bargis",
          "__typename": "Asset",
          "file": Object {
            "contentType": "image/jpeg",
            "details": Object {
              "image": Object {
                "height": 2828,
                "width": 4242,
              },
              "size": 3996687,
            },
            "fileName": "bargis.jpg",
            "url": "//images.ctfassets.net/abc/bargis/456/bargis.jpg",
          },
          "title": "Bargis",
        },
        "name": "Baz",
      },
      Object {
        "__id": "qux",
        "__typename": "Widget",
        "child": [Circular],
        "image": Object {
          "__id": "bargis",
          "__typename": "Asset",
          "file": Object {
            "contentType": "image/jpeg",
            "details": Object {
              "image": Object {
                "height": 2828,
                "width": 4242,
              },
              "size": 3996687,
            },
            "fileName": "bargis.jpg",
            "url": "//images.ctfassets.net/abc/bargis/456/bargis.jpg",
          },
          "title": "Bargis",
        },
        "name": "Qux",
      },
    ],
  },
  Object {
    "__id": "bar",
    "__typename": "Widget",
    "child": Object {
      "__id": "baz",
      "__typename": "Widget",
      "child": Object {
        "__id": "qux",
        "__typename": "Widget",
        "child": [Circular],
        "image": Object {
          "__id": "bargis",
          "__typename": "Asset",
          "file": Object {
            "contentType": "image/jpeg",
            "details": Object {
              "image": Object {
                "height": 2828,
                "width": 4242,
              },
              "size": 3996687,
            },
            "fileName": "bargis.jpg",
            "url": "//images.ctfassets.net/abc/bargis/456/bargis.jpg",
          },
          "title": "Bargis",
        },
        "name": "Qux",
      },
      "image": Object {
        "__id": "bargis",
        "__typename": "Asset",
        "file": Object {
          "contentType": "image/jpeg",
          "details": Object {
            "image": Object {
              "height": 2828,
              "width": 4242,
            },
            "size": 3996687,
          },
          "fileName": "bargis.jpg",
          "url": "//images.ctfassets.net/abc/bargis/456/bargis.jpg",
        },
        "title": "Bargis",
      },
      "name": "Baz",
    },
    "image": Object {
      "__id": "padasjoki",
      "__typename": "Asset",
      "file": Object {
        "contentType": "image/jpeg",
        "details": Object {
          "image": Object {
            "height": 3000,
            "width": 4000,
          },
          "size": 5080466,
        },
        "fileName": "padasjoki.jpg",
        "url": "//images.ctfassets.net/abc/padasjoki/123/padasjoki.jpg",
      },
      "title": "padasjoki",
    },
    "name": "Bar",
    "relatives": Array [
      Object {
        "__id": "foo",
        "__typename": "Widget",
        "child": [Circular],
        "description": "An English Foo",
        "image": Object {
          "__id": "bargis",
          "__typename": "Asset",
          "file": Object {
            "contentType": "image/jpeg",
            "details": Object {
              "image": Object {
                "height": 2828,
                "width": 4242,
              },
              "size": 3996687,
            },
            "fileName": "bargis.jpg",
            "url": "//images.ctfassets.net/abc/bargis/456/bargis.jpg",
          },
          "title": "Bargis",
        },
        "images": Array [
          Object {
            "__id": "bargis",
            "__typename": "Asset",
            "file": Object {
              "contentType": "image/jpeg",
              "details": Object {
                "image": Object {
                  "height": 2828,
                  "width": 4242,
                },
                "size": 3996687,
              },
              "fileName": "bargis.jpg",
              "url": "//images.ctfassets.net/abc/bargis/456/bargis.jpg",
            },
            "title": "Bargis",
          },
          Object {
            "__id": "padasjoki",
            "__typename": "Asset",
            "file": Object {
              "contentType": "image/jpeg",
              "details": Object {
                "image": Object {
                  "height": 3000,
                  "width": 4000,
                },
                "size": 5080466,
              },
              "fileName": "padasjoki.jpg",
              "url": "//images.ctfassets.net/abc/padasjoki/123/padasjoki.jpg",
            },
            "title": "padasjoki",
          },
        ],
        "name": "Foo",
        "relatives": Array [
          Object {
            "__id": "baz",
            "__typename": "Widget",
            "child": Object {
              "__id": "qux",
              "__typename": "Widget",
              "child": [Circular],
              "image": Object {
                "__id": "bargis",
                "__typename": "Asset",
                "file": Object {
                  "contentType": "image/jpeg",
                  "details": Object {
                    "image": Object {
                      "height": 2828,
                      "width": 4242,
                    },
                    "size": 3996687,
                  },
                  "fileName": "bargis.jpg",
                  "url": "//images.ctfassets.net/abc/bargis/456/bargis.jpg",
                },
                "title": "Bargis",
              },
              "name": "Qux",
            },
            "image": Object {
              "__id": "bargis",
              "__typename": "Asset",
              "file": Object {
                "contentType": "image/jpeg",
                "details": Object {
                  "image": Object {
                    "height": 2828,
                    "width": 4242,
                  },
                  "size": 3996687,
                },
                "fileName": "bargis.jpg",
                "url": "//images.ctfassets.net/abc/bargis/456/bargis.jpg",
              },
              "title": "Bargis",
            },
            "name": "Baz",
          },
          Object {
            "__id": "qux",
            "__typename": "Widget",
            "child": [Circular],
            "image": Object {
              "__id": "bargis",
              "__typename": "Asset",
              "file": Object {
                "contentType": "image/jpeg",
                "details": Object {
                  "image": Object {
                    "height": 2828,
                    "width": 4242,
                  },
                  "size": 3996687,
                },
                "fileName": "bargis.jpg",
                "url": "//images.ctfassets.net/abc/bargis/456/bargis.jpg",
              },
              "title": "Bargis",
            },
            "name": "Qux",
          },
        ],
      },
    ],
  },
  Object {
    "__id": "baz",
    "__typename": "Widget",
    "child": Object {
      "__id": "qux",
      "__typename": "Widget",
      "child": [Circular],
      "image": Object {
        "__id": "bargis",
        "__typename": "Asset",
        "file": Object {
          "contentType": "image/jpeg",
          "details": Object {
            "image": Object {
              "height": 2828,
              "width": 4242,
            },
            "size": 3996687,
          },
          "fileName": "bargis.jpg",
          "url": "//images.ctfassets.net/abc/bargis/456/bargis.jpg",
        },
        "title": "Bargis",
      },
      "name": "Qux",
    },
    "image": Object {
      "__id": "bargis",
      "__typename": "Asset",
      "file": Object {
        "contentType": "image/jpeg",
        "details": Object {
          "image": Object {
            "height": 2828,
            "width": 4242,
          },
          "size": 3996687,
        },
        "fileName": "bargis.jpg",
        "url": "//images.ctfassets.net/abc/bargis/456/bargis.jpg",
      },
      "title": "Bargis",
    },
    "name": "Baz",
  },
  Object {
    "__id": "qux",
    "__typename": "Widget",
    "child": [Circular],
    "image": Object {
      "__id": "bargis",
      "__typename": "Asset",
      "file": Object {
        "contentType": "image/jpeg",
        "details": Object {
          "image": Object {
            "height": 2828,
            "width": 4242,
          },
          "size": 3996687,
        },
        "fileName": "bargis.jpg",
        "url": "//images.ctfassets.net/abc/bargis/456/bargis.jpg",
      },
      "title": "Bargis",
    },
    "name": "Qux",
  },
  Object {
    "__id": "doodah",
    "__typename": "Gadget",
    "name": "Doodah",
  },
]
`;

exports[`ContentfulStore store should load 2`] = `
Array [
  Object {
    "__id": "bargis",
    "__typename": "Asset",
    "file": Object {
      "contentType": "image/jpeg",
      "details": Object {
        "image": Object {
          "height": 2828,
          "width": 4242,
        },
        "size": 3996687,
      },
      "fileName": "bargis.jpg",
      "url": "//images.ctfassets.net/abc/bargis/456/bargis.jpg",
    },
    "title": "Bargis",
  },
  Object {
    "__id": "padasjoki",
    "__typename": "Asset",
    "file": Object {
      "contentType": "image/jpeg",
      "details": Object {
        "image": Object {
          "height": 3000,
          "width": 4000,
        },
        "size": 5080466,
      },
      "fileName": "padasjoki.jpg",
      "url": "//images.ctfassets.net/abc/padasjoki/123/padasjoki.jpg",
    },
    "title": "padasjoki",
  },
]
`;

exports[`ContentfulStore store should sync 1`] = `
Array [
  Object {
    "__id": "foo",
    "__typename": "Widget",
    "child": Object {
      "__id": "qux",
      "__typename": "Widget",
      "child": [Circular],
      "image": undefined,
      "name": "Qux",
    },
    "description": "An updated Foo",
    "images": Array [
      Object {
        "__id": "padasjoki",
        "__typename": "Asset",
        "file": Object {
          "contentType": "image/jpeg",
          "details": Object {
            "image": Object {
              "height": 3000,
              "width": 4000,
            },
            "size": 5080466,
          },
          "fileName": "padasjoki.jpg",
          "url": "//images.ctfassets.net/abc/padasjoki/789/padasjoki.jpg",
        },
        "title": "Padasjoki Updated",
      },
    ],
    "name": "Foo",
    "relatives": Array [
      Object {
        "__id": "baz",
        "__typename": "Widget",
        "child": Object {
          "__id": "qux",
          "__typename": "Widget",
          "child": [Circular],
          "image": undefined,
          "name": "Qux",
        },
        "image": undefined,
        "name": "Baz",
      },
      Object {
        "__id": "qux",
        "__typename": "Widget",
        "child": [Circular],
        "image": undefined,
        "name": "Qux",
      },
    ],
  },
  Object {
    "__id": "baz",
    "__typename": "Widget",
    "child": Object {
      "__id": "qux",
      "__typename": "Widget",
      "child": [Circular],
      "image": undefined,
      "name": "Qux",
    },
    "image": undefined,
    "name": "Baz",
  },
  Object {
    "__id": "qux",
    "__typename": "Widget",
    "child": [Circular],
    "image": undefined,
    "name": "Qux",
  },
  Object {
    "__id": "doodah",
    "__typename": "Gadget",
    "name": "Doodah",
  },
]
`;

exports[`ContentfulStore store should sync 2`] = `
Array [
  Object {
    "__id": "padasjoki",
    "__typename": "Asset",
    "file": Object {
      "contentType": "image/jpeg",
      "details": Object {
        "image": Object {
          "height": 3000,
          "width": 4000,
        },
        "size": 5080466,
      },
      "fileName": "padasjoki.jpg",
      "url": "//images.ctfassets.net/abc/padasjoki/789/padasjoki.jpg",
    },
    "title": "Padasjoki Updated",
  },
]
`;
